client/server/pages/index.js

```diff
@@ -745,22 +745,25 @@ export default function pages() {
 		}
 	);
 
-	app.get( '/browsehappy', ( req, res ) => {
-		// We only want to allow a redirect to calypso routes, so we check that
-		// the query param was generated by the same origin.
-		const protocol = process.env.PROTOCOL || config( 'protocol' );
-		const hostname = process.env.HOST || config( 'hostname' );
-		const port = process.env.PORT || config( 'port' );
-		const serverURL = `${ protocol }://${ hostname }${ port ? ':' + port : '' }`;
-
-		// Matches URLs which begin with exactly "https://wordpress.com/" (or other calypso domains.)
-		const baseURLMatcher = new RegExp( `^${ serverURL }/` );
-
-		const from = req.query.from;
-		let redirectLocation = '/';
-		if ( from && baseURLMatcher.test( from ) ) {
-			redirectLocation = from.replace( baseURLMatcher, '/' );
+	function validateRedirect( req, url ) {
+		if ( ! url ) {
+			return false;
+		}
+
+		try {
+			const serverOrigin = req.protocol + '://' + req.host;
+			return new URL( url, serverOrigin ).origin === serverOrigin;
+		} catch {
+			// if parsing the URL fails, it is not valid
+			return false;
 		}
+	}
+
+	app.get( '/browsehappy', ( req, res ) => {
+		// We only want to allow a redirect to Calypso routes, so we check that
+		// the `from` query param has the same origin.
+		const { from } = req.query;
+		const redirectLocation = from && validateRedirect( req, from ) ? from : '/';
```

Comment:
In `validatateRedirect` we check for `! url` so can the `from &&` part be omitted?

Reply:
That's true, but as `from` is optional, it's probably not bad to say "if `from` is present and it's valid, then ..." anyway.